Fixed crash when reporting successful login if "Session Log" is enabled.

Enmk opened the pull request:
Changelog category (leave one):
- Bug Fix (user-visible misbehaviour in official stable or prestable release)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed crash when reporting successful login in some cases.
...

Closes: #35714

Enmk committed:
Fixed crash when reporting successful login.

tavplubix commented on src/Interpreters/SessionLog.cpp:
```diff
@@ -212,8 +212,8 @@ void SessionLog::addLoginSuccess(const UUID & auth_id, std::optional<String> ses
     DB::SessionLogElement log_entry(auth_id, SESSION_LOGIN_SUCCESS);
     log_entry.client_info = client_info;
 
+    if (const auto user = login_context.getUser())
```
Why can it be nullptr? Seems like we cannot have a session without user.

alexey-milovidov replied:
+1, if login sucessful, we should expect the user exists.

Enmk replied:
But this somehow happened in the test.

`Context::getUser()` is equivalent to `context.getAccess()->getUser()`:
https://github.com/ClickHouse/ClickHouse/blob/master/src/Interpreters/Context.cpp#L750

In `Session::makeQueryContextImpl` we check that `user` is non-null before invoking `SessionLog::addLoginSuccess`: https://github.com/ClickHouse/ClickHouse/blob/master/src/Interpreters/Session.cpp#L484

Yet somehow it became null in `SessionLog::addLoginSuccess` in the [test](https://github.com/ClickHouse/ClickHouse/issues/35714).

The context we are working on is local, and can't be shared between multiple threads, but `ContextAccess` is shared between contexts (e.g. original session context). It's `ContextAccess::user` and it can be changed and set to null in a change-handling callback:
https://github.com/ClickHouse/ClickHouse/blob/master/src/Access/ContextAccess.cpp#L163

tavplubix replied:
Can it be a race condition with `DROP USER`, for example? Need to check server logs. In this case on successful login we should save shared pointer to `user` somewhere to be able to write meaningful row into session log. Imagine that you are checking login attempts in session log on your server and see some successful logins without information about user... I suppose that it will really scare you, because it looks like someone has hacked your server.

tavplubix replied:
I didn't find the corresponding `DROP USER` query in server logs, so it's still not clear what happened.

Enmk committed:
Explicitly passing a user object to create a LogIn event

alexey-milovidov commented:
Let'a add a test.

Enmk committed:
Test that tries to reproduce a race condition

Enmk committed:
Made test less flakky

tavplubix commented:
```
2022-04-01 01:40:58 02242_delete_user_race:                                                 [ OK ] 5.21 sec.
2022-04-01 01:41:03 02242_delete_user_race:                                                 [ OK ] 4.95 sec.
2022-04-01 01:41:08 02242_delete_user_race:                                                 [ OK ] 5.15 sec.
2022-04-01 01:51:09 02242_delete_user_race:                                                 [ FAIL ] 600.47 sec. - Timeout!
2022-04-01 01:51:09 Settings used in the test: --max_insert_threads=10 --group_by_two_level_threshold=100000 --group_by_two_level_threshold_bytes=50000000 --distributed_aggregation_memory_efficient=0 --fsync_metadata=0 --priority=4 --output_format_parallel_formatting=0 --input_format_parallel_parsing=1 --min_chunk_bytes_for_parallel_parsing=17960349 --max_read_buffer_size=868583 --prefer_localhost_replica=0 --max_block_size=41584 --max_threads=5
2022-04-01 01:51:09 
2022-04-01 01:51:09 Database: test_l3v6if
2022-04-01 01:51:15 02242_delete_user_race:                                                 [ OK ] 5.73 sec.
2022-04-01 01:51:21 02242_delete_user_race:                                                 [ OK ] 5.31 sec.
2022-04-01 01:51:27 02242_delete_user_race:                                                 [ OK ] 5.50 sec.
2022-04-01 01:51:32 02242_delete_user_race:                                                 [ OK ] 5.33 sec.
```

Looks like a deadlock. I don't think that reducing number of iterations from 1000 to 100 will really help, we should investigate why this timeout happened.

azat commented on tests/queries/0_stateless/02242_delete_user_race.sh:
```diff
@@ -0,0 +1,78 @@
+function set_role()
+{
+    local i
+    for (( i = 0; i < REPEAT; ++i ))
```
This is not correct usage of `clickhouse_client_loop_timeout`, especially with 0.1 timeout, since it may send INT to this function before the client will execute any query, and if you will take at the logs you will see for example that there is zero `SET ROLE` queries.

azat commented:
>Looks like a deadlock. I don't think that reducing number of iterations from 1000 to 100 will really help, we should investigate why this timeout happened.

I've looked at the logs and they does not have any queries to clickhouse:

```sh
$ pigz -cd clickhouse-server.log.gz | grep -F -f <(clickhouse-local --file query_log.tsv.gz --input-format TSVWithNamesAndTypes -q "select distinct query_id from table where current_database = 'test_l3v6if'")
2022.04.01 01:41:09.875566 [ 385 ] {34dc998d-a7ad-4fb0-bc9d-4529e21b08b6} <Debug> executeQuery: (from [::1]:43106) (comment: 02242_delete_user_race.sh) DROP ROLE IF EXISTS test_role_02242;
2022.04.01 01:41:09.876314 [ 385 ] {34dc998d-a7ad-4fb0-bc9d-4529e21b08b6} <Trace> ContextAccess (default): Access granted: DROP ROLE ON *.*
2022.04.01 01:41:09.877758 [ 385 ] {34dc998d-a7ad-4fb0-bc9d-4529e21b08b6} <Debug> MemoryTracker: Peak memory usage (for query): 0.00 B.
2022.04.01 01:41:09.880693 [ 385 ] {8d6fb637-2cb1-49c1-ab92-b15fa7d0a75a} <Debug> executeQuery: (from [::1]:43106) (comment: 02242_delete_user_race.sh) CREATE ROLE test_role_02242;
2022.04.01 01:41:09.881434 [ 385 ] {8d6fb637-2cb1-49c1-ab92-b15fa7d0a75a} <Trace> ContextAccess (default): Access granted: CREATE ROLE ON *.*
2022.04.01 01:41:09.883165 [ 385 ] {8d6fb637-2cb1-49c1-ab92-b15fa7d0a75a} <Debug> MemoryTracker: Peak memory usage (for query): 3.00 MiB.
```

So it is something with the test itself, maybe some issue with `wait` in `bash` (something related to `wait_for: No record of process` error), this test requires some logging to investigate this.

Enmk committed:
Fixed flaky test

Enmk committed:
Fixed test

alexey-milovidov commented:
@Enmk we still observe timeouts in tests.

tavplubix commented:
Actually it's segfault, not timeout:
```
2022.04.28 19:03:29.481245 [ 153165 ] {} <Fatal> BaseDaemon: ########################################
2022.04.28 19:03:29.481304 [ 153165 ] {} <Fatal> BaseDaemon: (version 22.5.1.1, build id: 9203DFA55512D6D8) (from thread 152925) (no query) Received signal Segmentation fault (11)
2022.04.28 19:03:29.481337 [ 153165 ] {} <Fatal> BaseDaemon: Address: 0x138 Access: read. Address not mapped to object.
2022.04.28 19:03:29.481353 [ 153165 ] {} <Fatal> BaseDaemon: Stack trace: 0x154fbe44 0x15cf31a8 0x16273fba 0x17008ad5 0x17017579 0x1b3e358f 0x1b3e59e1 0x1b5ac429 0x1b5a9780 0x7f8bb89e5609 0x7f8bb890a163
2022.04.28 19:03:29.481404 [ 153165 ] {} <Fatal> BaseDaemon: 2. DB::GrantedRoles::findGranted(DB::RolesOrUsersSet const&) const @ 0x154fbe44 in /usr/bin/clickhouse
2022.04.28 19:03:29.481429 [ 153165 ] {} <Fatal> BaseDaemon: 3. DB::Context::setUser(StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag> const&) @ 0x15cf31a8 in /usr/bin/clickhouse
2022.04.28 19:03:29.481449 [ 153165 ] {} <Fatal> BaseDaemon: 4. DB::Session::makeSessionContext() @ 0x16273fba in /usr/bin/clickhouse
2022.04.28 19:03:29.481469 [ 153165 ] {} <Fatal> BaseDaemon: 5. DB::TCPHandler::runImpl() @ 0x17008ad5 in /usr/bin/clickhouse
2022.04.28 19:03:29.481496 [ 153165 ] {} <Fatal> BaseDaemon: 6. DB::TCPHandler::run() @ 0x17017579 in /usr/bin/clickhouse
2022.04.28 19:03:29.481514 [ 153165 ] {} <Fatal> BaseDaemon: 7. Poco::Net::TCPServerConnection::start() @ 0x1b3e358f in /usr/bin/clickhouse
2022.04.28 19:03:29.481530 [ 153165 ] {} <Fatal> BaseDaemon: 8. Poco::Net::TCPServerDispatcher::run() @ 0x1b3e59e1 in /usr/bin/clickhouse
2022.04.28 19:03:29.481548 [ 153165 ] {} <Fatal> BaseDaemon: 9. Poco::PooledThread::run() @ 0x1b5ac429 in /usr/bin/clickhouse
2022.04.28 19:03:29.481567 [ 153165 ] {} <Fatal> BaseDaemon: 10. Poco::ThreadImpl::runnableEntry(void*) @ 0x1b5a9780 in /usr/bin/clickhouse
2022.04.28 19:03:29.481581 [ 153165 ] {} <Fatal> BaseDaemon: 11. ? @ 0x7f8bb89e5609 in ?
2022.04.28 19:03:29.481594 [ 153165 ] {} <Fatal> BaseDaemon: 12. clone @ 0x7f8bb890a163 in ?
2022.04.28 19:03:30.869445 [ 153165 ] {} <Fatal> BaseDaemon: Calculated checksum of the binary: F40ACC981B80710C79AC778284059B0B. There is no information about the reference checksum.
2022.04.28 19:03:49.892777 [ 387 ] {} <Fatal> Application: Child process was terminated by signal 11.
```

Enmk committed:
Fixed issues with concurrent access to User object

ContextAccess might be updated from another thread, nullifying UserPtr.
Can happen if user was dropped in concurrent query.

This could crash server under certain conditions.

alexey-milovidov commented:
As nothing was fixed in this PR and it is abandoned, let's close.
In the meantime I've added deprecation warnings around session_log,
so it will be removed in subsequent releases.

Enmk commented:
Well, the original issue in SessionLog was fixed, other issues discovered by tests in this PR were addressed in another PR: https://github.com/ClickHouse/ClickHouse/pull/36935

Enmk committed:
Merge branch 'master' into fix_session_log_crash

tavplubix commented:
But we cannot merge this PR because tests fail. So we need to merge #36935 first, then merge master into this branch and restart tests in this PR.

tavplubix commented on src/Interpreters/Session.cpp:
```diff
@@ -439,11 +439,12 @@ ContextMutablePtr Session::makeQueryContextImpl(const ClientInfo * client_info_t
               from_session_context ? "session" : "global",
               user_id ? toString(*user_id) : "<EMPTY>",
-              query_context->getUser() ? query_context->getUser()->getName() : "<NOT SET>");
+              query_context_user ? query_context_user->getName() : "<NOT SET>");
```
Messages like `Creating query context from session context, user_id: <NOT SET>` in logs look creepy (because session cannot be created without a user AFAIK). We should avoid such messages. And especially we should avoid messages like `Authenticated with global context as user <EMPTY>`. See also #30370

Enmk replied:
Well, first, this is a debug message (hence not really intended for the general public). Second, it just states how things are. But Ok, I removed the message if you don't like it.

tavplubix replied:
> Well, first, this is a debug message (hence not really intended for the general public)

Default log level in clickhouse is `trace` (yep, that's weird, but it works this way historically), it includes debug messages as well. We have `test` log level for messages that not intended for the general public.

> Second, it just states how things are. But Ok, I removed the message if you don't like it.

The problem is that things should not be like this and that's what I actually don't like. We should either save `UserPtr` somewhere to be able to print meaningful messages or throw exception if user is required but it's nullptr. Maybe it's ok to throw exception later, but at least comment is required in this case. Now this code looks like we may allow some unknown user to start a session without authentication.

And what about `Authenticated with global context as user <EMPTY>`? It's much more creepy, because it makes you think that someone has hacked your server. (and the same question about other messages added in #30370, I wouldn't merge it if I was a reviewer).

tavplubix replied:
Btw, I've just found these WTF-messages in server log on a production cluster:
```
2022.05.10 11:43:16.938357 [ 414 ] {} <Debug> TCP-Session: a0e10978-38c5-496c-bfe1-b5b270cb169a Destroying unnamed session of user <EMPTY>
2022.05.10 11:43:17.790144 [ 414 ] {} <Information> TCPHandler: Client has not sent any data.
2022.05.10 11:43:17.790156 [ 414 ] {} <Debug> TCPHandler: Done processing connection.
2022.05.10 11:43:17.790174 [ 414 ] {} <Debug> TCP-Session: a9eb1685-eae7-45ce-948e-3f72fa7faab8 Destroying unnamed session of user <EMPTY>
2022.05.10 11:43:17.993485 [ 414 ] {} <Information> TCPHandler: Client has not sent any data.
2022.05.10 11:43:17.993496 [ 414 ] {} <Debug> TCPHandler: Done processing connection.
2022.05.10 11:43:17.993512 [ 414 ] {} <Debug> TCP-Session: 642b742a-3037-4293-a6b5-4fe1cf14541a Destroying unnamed session of user <EMPTY>
2022.05.10 11:43:18.476784 [ 414 ] {} <Information> TCPHandler: Client has not sent any data.
2022.05.10 11:43:18.476795 [ 414 ] {} <Debug> TCPHandler: Done processing connection.
```

Enmk committed:
No debug messge if user not set

qoega commented on tests/queries/0_stateless/02242_delete_user_race.sh:
```diff
@@ -0,0 +1,62 @@
+#!/usr/bin/env bash
+# Tags: race, no-fasttest, no-parallel
```
Here

Enmk committed:
no-backward-compatibility-check tag to the test

Enmk committed:
Added checks to validate that authentication happens before creating query or session context;

Updated logging of user_id, removed unnecessary precautions
Explicitly stating in Session d-tor if user was not authenticated

Enmk committed:
Less obscure message on destroying Session

tavplubix commented:
Integration tests:
```
2022.05.13 14:07:59.390484 [ 10 ] {} <Fatal> : Logical error: 'Session context must be created after authentication'.
2022.05.13 14:07:59.390919 [ 228 ] {} <Fatal> BaseDaemon: ########################################
2022.05.13 14:07:59.390988 [ 228 ] {} <Fatal> BaseDaemon: (version 22.5.1.1, build id: B1B5F197544B0200) (from thread 10) (no query) Received signal Aborted (6)
2022.05.13 14:07:59.391039 [ 228 ] {} <Fatal> BaseDaemon: 
2022.05.13 14:07:59.391074 [ 228 ] {} <Fatal> BaseDaemon: Stack trace: 0x7fc7f749803b 0x7fc7f7477859 0xd26a98f 0xd26afee 0x2b8fe326 0x2dd71ddc 0x2dd5ea32 0x2dd4f10b 0x2dd79700 0x387e8a0f 0x387e9781 0x38cd4126 0x38ccd3da 0x7fc7f764f609 0x7fc7f7574163
2022.05.13 14:07:59.391116 [ 228 ] {} <Fatal> BaseDaemon: 3. gsignal @ 0x7fc7f749803b in ?
2022.05.13 14:07:59.391135 [ 228 ] {} <Fatal> BaseDaemon: 4. abort @ 0x7fc7f7477859 in ?
2022.05.13 14:07:59.439293 [ 228 ] {} <Fatal> BaseDaemon: 5. ./build_docker/../src/Common/Exception.cpp:0: DB::handle_error_code(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool, std::__1::vector<void*, std::__1::allocator<void*> > const&) @ 0xd26a98f in /usr/bin/clickhouse
2022.05.13 14:07:59.487284 [ 228 ] {} <Fatal> BaseDaemon: 6.1. inlined from ./build_docker/../contrib/libcxx/include/vector:547: std::__1::vector<void*, std::__1::allocator<void*> >::data() const
2022.05.13 14:07:59.487342 [ 228 ] {} <Fatal> BaseDaemon: 6.2. inlined from ../contrib/libcxx/include/vector:731: std::__1::vector<void*, std::__1::allocator<void*> >::__annotate_delete() const
2022.05.13 14:07:59.487360 [ 228 ] {} <Fatal> BaseDaemon: 6.3. inlined from ../contrib/libcxx/include/vector:394: ~vector
2022.05.13 14:07:59.487399 [ 228 ] {} <Fatal> BaseDaemon: 6. ../src/Common/Exception.cpp:59: DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0xd26afee in /usr/bin/clickhouse
2022.05.13 14:07:59.528933 [ 228 ] {} <Fatal> BaseDaemon: 7. ./build_docker/../src/Interpreters/Session.cpp:439: DB::Session::makeQueryContextImpl(DB::ClientInfo const*, DB::ClientInfo*) const @ 0x2b8fe326 in /usr/bin/clickhouse
2022.05.13 14:07:59.621003 [ 228 ] {} <Fatal> BaseDaemon: 8. ./build_docker/../src/Server/TCPHandler.cpp:0: DB::TCPHandler::receiveQuery() @ 0x2dd71ddc in /usr/bin/clickhouse
2022.05.13 14:07:59.729693 [ 228 ] {} <Fatal> BaseDaemon: 9. ./build_docker/../src/Server/TCPHandler.cpp:0: DB::TCPHandler::receivePacket() @ 0x2dd5ea32 in /usr/bin/clickhouse
2022.05.13 14:07:59.812245 [ 228 ] {} <Fatal> BaseDaemon: 10. ./build_docker/../src/Server/TCPHandler.cpp:0: DB::TCPHandler::runImpl() @ 0x2dd4f10b in /usr/bin/clickhouse
2022.05.13 14:07:59.955042 [ 228 ] {} <Fatal> BaseDaemon: 11. ./build_docker/../src/Server/TCPHandler.cpp:1783: DB::TCPHandler::run() @ 0x2dd79700 in /usr/bin/clickhouse
2022.05.13 14:07:59.958378 [ 228 ] {} <Fatal> BaseDaemon: 12. ./build_docker/../contrib/poco/Net/src/TCPServerConnection.cpp:57: Poco::Net::TCPServerConnection::start() @ 0x387e8a0f in /usr/bin/clickhouse
2022.05.13 14:07:59.967522 [ 228 ] {} <Fatal> BaseDaemon: 13.1. inlined from ./build_docker/../contrib/libcxx/include/__memory/unique_ptr.h:54: std::__1::default_delete<Poco::Net::TCPServerConnection>::operator()(Poco::Net::TCPServerConnection*) const
2022.05.13 14:07:59.967573 [ 228 ] {} <Fatal> BaseDaemon: 13.2. inlined from ../contrib/libcxx/include/__memory/unique_ptr.h:315: std::__1::unique_ptr<Poco::Net::TCPServerConnection, std::__1::default_delete<Poco::Net::TCPServerConnection> >::reset(Poco::Net::TCPServerConnection*)
2022.05.13 14:07:59.967595 [ 228 ] {} <Fatal> BaseDaemon: 13.3. inlined from ../contrib/libcxx/include/__memory/unique_ptr.h:269: ~unique_ptr
2022.05.13 14:07:59.967612 [ 228 ] {} <Fatal> BaseDaemon: 13. ../contrib/poco/Net/src/TCPServerDispatcher.cpp:116: Poco::Net::TCPServerDispatcher::run() @ 0x387e9781 in /usr/bin/clickhouse
2022.05.13 14:07:59.977422 [ 228 ] {} <Fatal> BaseDaemon: 14. ./build_docker/../contrib/poco/Foundation/src/ThreadPool.cpp:0: Poco::PooledThread::run() @ 0x38cd4126 in /usr/bin/clickhouse
2022.05.13 14:07:59.986786 [ 228 ] {} <Fatal> BaseDaemon: 15. ./build_docker/../contrib/poco/Foundation/src/Thread_POSIX.cpp:0: Poco::ThreadImpl::runnableEntry(void*) @ 0x38ccd3da in /usr/bin/clickhouse
2022.05.13 14:07:59.986826 [ 228 ] {} <Fatal> BaseDaemon: 16. ? @ 0x7fc7f764f609 in ?
2022.05.13 14:07:59.986842 [ 228 ] {} <Fatal> BaseDaemon: 17. clone @ 0x7fc7f7574163 in ?
```

Enmk committed:
Removed an unnecessary check and extra logging

Enmk commented:
> Integration tests:
> 
> ```
> 2022.05.13 14:07:59.390484 [ 10 ] {} <Fatal> : Logical error: 'Session context must be created after authentication'.
> ```

Yep, turns out session can be used in very funny ways... I removed the exception, and removed the logging
